README.md
# aws-sso-auth

AWS SSO is supported by the aws cli v2, but it does not provide credentials for using the AWS SDK (as far as I can work out).

This application leverages the aws cli support to provide a simple way to get credentials for your normal day to day aws usage.

## AWS CLI v2

Only version 2 of the CLI supports SSO. You will need to install this and make sure it is being used.

The Linux installation guide is here:

https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2-linux.html

## Installation

Oneline:
```
$ bash <(curl "https://raw.githubusercontent.com/isotoma/aws-sso-auth/main/install.sh")
```

Or, if you don't like running scripts from the internet:
- download an executable from https://github.com/isotoma/aws-sso-auth/releases
- make it executable with `chmod a+x`
- put it somewhere that is on your `$PATH`

Or, if you don't trust those executables:
- Checkout this repository
- Run `npm run package`
- Take one of the executables from `./dist/`
- put it somewhere that is on your `$PATH`

Or, if you don't trust executables made by pkg:
- Checkout this repository
- Run `npm run build`
- Alias `node /path/to/repo/build/bin.js` to `aws-sso-auth`

## Usage

You will need to have a current SSO session with the AWS CLI. Before using the AWS CLI for SSO you need to configure it with `aws sso configure`. By default, `aws-sso-auth` the profile called `default` for your sso login. Change this with the `--profile` flag if needed.

Run:

```
aws-sso-auth
```

This checks for temporary credentials in `~/.aws/sso/cache/`, then **overwrites** `~/.aws/credentials` with temporary credentials retrieved using `aws sso get-role-credentials`.

## Usage with `credentials_process`

Rather than overwriting `~/.aws/credentials`, provided the SDK/program you need to read AWS credentials can make use of the `credentials_process` option (see https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-sourcing-external.html), you can instead put the following in a profile that also has the `sso_...` configuration in `~/.aws/config`:

```
credential_process = /usr/local/bin/aws-sso-auth credentials-process
```
(or a different path, if you've installed elsewhere - note it needs to be an absolute path)

This way, whenever anything needs AWS credentials, it will call that command. This caches credentials in `~/.aws-sso-auth-credentials.json`, rather than touching anything in `~/.aws`.

## Finding version of an executable

Inspect the version of an executable with:
```
aws-sso-auth version
```

## Development

```
npm install
```

Change code, then, to compile and run:

```
npm run build && node build/bin.js
```

and run the tests with:

```
npm run test
```

You can build standalone executables with:

```
npm run package
```

and executables are produced in `./dist/`.


CHANGELOG.md
### Changelog

#### [2.9.0](https://github.com/isotoma/aws-sso-auth/compare/2.8.1...2.9.0)

- Display confirmation code [`#67`](https://github.com/isotoma/aws-sso-auth/pull/67)

#### [2.8.1](https://github.com/isotoma/aws-sso-auth/compare/2.8.0...2.8.1)

> 15 August 2023

- Remove json-schema [`#65`](https://github.com/isotoma/aws-sso-auth/pull/65)
- Update jest dependencies [`#64`](https://github.com/isotoma/aws-sso-auth/pull/64)
- Update linting dependencies [`#63`](https://github.com/isotoma/aws-sso-auth/pull/63)
- Update production dependencies [`#62`](https://github.com/isotoma/aws-sso-auth/pull/62)
- Update non-linting, non-jest dev dependencies [`#61`](https://github.com/isotoma/aws-sso-auth/pull/61)
- Package updates [`#60`](https://github.com/isotoma/aws-sso-auth/pull/60)

#### [2.8.0](https://github.com/isotoma/aws-sso-auth/compare/2.7.0...2.8.0)

> 15 August 2023

- Issue 58: credentials expiry warning and workaround [`#59`](https://github.com/isotoma/aws-sso-auth/pull/59)

#### [2.7.0](https://github.com/isotoma/aws-sso-auth/compare/2.6.0...2.7.0)

> 14 June 2023

- Use node 18 [`#57`](https://github.com/isotoma/aws-sso-auth/pull/57)

#### [2.6.0](https://github.com/isotoma/aws-sso-auth/compare/2.5.0...2.6.0)

> 14 June 2023

- Passing --verbose also prints the version [`#56`](https://github.com/isotoma/aws-sso-auth/pull/56)

#### [2.5.0](https://github.com/isotoma/aws-sso-auth/compare/2.4.1...2.5.0)

> 16 March 2023

- Add --force and --skip-expiry-check options [`#51`](https://github.com/isotoma/aws-sso-auth/pull/51)
- SonarCloud: ignore duplication in tests [`#53`](https://github.com/isotoma/aws-sso-auth/pull/53)
- Fix CI [`#52`](https://github.com/isotoma/aws-sso-auth/pull/52)
- Upgrade ansi-regex [`#47`](https://github.com/isotoma/aws-sso-auth/pull/47)
- Upgrade json-schema [`#46`](https://github.com/isotoma/aws-sso-auth/pull/46)
- Bump simple-get from 3.1.0 to 3.1.1 [`#45`](https://github.com/isotoma/aws-sso-auth/pull/45)
- Bump node-fetch from 2.6.1 to 2.6.7 [`#44`](https://github.com/isotoma/aws-sso-auth/pull/44)
- Bump tmpl from 1.0.4 to 1.0.5 [`#43`](https://github.com/isotoma/aws-sso-auth/pull/43)
- Bump ws from 7.4.3 to 7.5.7 [`#42`](https://github.com/isotoma/aws-sso-auth/pull/42)
- Bump glob-parent from 5.1.1 to 5.1.2 [`#41`](https://github.com/isotoma/aws-sso-auth/pull/41)
- Bump axios from 0.21.1 to 0.21.4 [`#40`](https://github.com/isotoma/aws-sso-auth/pull/40)
- Bump path-parse from 1.0.6 to 1.0.7 [`#39`](https://github.com/isotoma/aws-sso-auth/pull/39)

#### [2.4.1](https://github.com/isotoma/aws-sso-auth/compare/2.4.0...2.4.1)

> 14 February 2022

- Bump follow-redirects from 1.14.1 to 1.14.8 [`#38`](https://github.com/isotoma/aws-sso-auth/pull/38)
- Use npm v7 [`#37`](https://github.com/isotoma/aws-sso-auth/pull/37)

#### [2.4.0](https://github.com/isotoma/aws-sso-auth/compare/2.3.1...2.4.0)

> 11 May 2021

- Update pkg to 5.1.0 [`#36`](https://github.com/isotoma/aws-sso-auth/pull/36)
- Bump hosted-git-info from 2.8.8 to 2.8.9 [`#35`](https://github.com/isotoma/aws-sso-auth/pull/35)
- Bump lodash from 4.17.20 to 4.17.21 [`#34`](https://github.com/isotoma/aws-sso-auth/pull/34)
- Docs: update for master-&gt;main default branch name [`20b012b`](https://github.com/isotoma/aws-sso-auth/commit/20b012b319c43b5c8f5205f885afa1fa1ef3c831)

#### [2.3.1](https://github.com/isotoma/aws-sso-auth/compare/2.3.0...2.3.1)

> 1 April 2021

- Bump y18n from 4.0.0 to 4.0.1 [`#33`](https://github.com/isotoma/aws-sso-auth/pull/33)

#### [2.3.0](https://github.com/isotoma/aws-sso-auth/compare/2.2.7...2.3.0)

> 16 February 2021

- Use node v14 [`#31`](https://github.com/isotoma/aws-sso-auth/pull/31)

#### [2.2.7](https://github.com/isotoma/aws-sso-auth/compare/2.2.6...2.2.7)

> 16 February 2021

- Tidying for sonarcloud [`#30`](https://github.com/isotoma/aws-sso-auth/pull/30)
- Upgrade dev dependencies [`#29`](https://github.com/isotoma/aws-sso-auth/pull/29)

#### [2.2.6](https://github.com/isotoma/aws-sso-auth/compare/2.2.5...2.2.6)

> 23 December 2020

- Bump ini from 1.3.5 to 1.3.6 [`#28`](https://github.com/isotoma/aws-sso-auth/pull/28)

#### [2.2.5](https://github.com/isotoma/aws-sso-auth/compare/2.2.4...2.2.5)

> 10 November 2020

- Bump node-fetch from 2.6.0 to 2.6.1 [`#25`](https://github.com/isotoma/aws-sso-auth/pull/25)
- Tests: rename content variable to avoid name collision [`#27`](https://github.com/isotoma/aws-sso-auth/pull/27)
- Rename yargs object in parsing failure handler [`#26`](https://github.com/isotoma/aws-sso-auth/pull/26)

#### [2.2.4](https://github.com/isotoma/aws-sso-auth/compare/2.2.3...2.2.4)

> 6 August 2020

- Session token expiry workaround (#23) [`#24`](https://github.com/isotoma/aws-sso-auth/pull/24)

#### [2.2.3](https://github.com/isotoma/aws-sso-auth/compare/2.2.2...2.2.3)

> 24 July 2020

- Bump lodash from 4.17.15 to 4.17.19 [`#22`](https://github.com/isotoma/aws-sso-auth/pull/22)

#### [2.2.2](https://github.com/isotoma/aws-sso-auth/compare/2.2.1...2.2.2)

> 24 June 2020

- Fix docs for credential_process [`882df93`](https://github.com/isotoma/aws-sso-auth/commit/882df93fa13b49cbf929e24b751ffb0fac06cde7)

#### [2.2.1](https://github.com/isotoma/aws-sso-auth/compare/2.2.0...2.2.1)

> 24 June 2020

- Update docs for addition of --profile flag [`56c39eb`](https://github.com/isotoma/aws-sso-auth/commit/56c39ebb6911b6e4abe780d7e9074399ac65e3af)
- Update docs to be more helpful for setting credential_process [`38e1f1d`](https://github.com/isotoma/aws-sso-auth/commit/38e1f1df6de2248d8709b56ac5fa5e5602d8be44)
- Make install confirmation prompt clearer [`c4fa01a`](https://github.com/isotoma/aws-sso-auth/commit/c4fa01a52d760b5f4f3af2c564e89ad9d8ff7f6b)

#### [2.2.0](https://github.com/isotoma/aws-sso-auth/compare/2.1.1...2.2.0)

> 24 June 2020

- Configurable profile name (#20) [`#21`](https://github.com/isotoma/aws-sso-auth/pull/21)
- Owner-only access permissions on created files (#14) [`#19`](https://github.com/isotoma/aws-sso-auth/pull/19)

src/main.ts
import util from 'util';
import { exec, spawn } from 'child_process';

import yargs from 'yargs';

import { ArgumentsError, NoCachedCredentialsError, BadAWSCLIVersionError, MisbehavingExpiryDateError } from './errors';
import { checkCLIVersion } from './checkCLIVersion';
import { getVersionNumber } from './getVersion';
import { findLatestCacheFile, SSOCacheToken, deleteCredentialsAndCaches } from './cache';
import { findSSOConfigFromAWSConfig, SSOConfigOptions } from './ssoConfig';
import { parseRoleCredentialsOutput, writeCredentialsFile, writeCredentialsCacheFile, readCredentialsCacheFile, printCredentials } from './roleCredentials';
import { hasKey } from './utils';

const execPromise = util.promisify(exec);

const jsSdkExpiryWindowSeconds = 15 * 60;

interface RunProps {
    verbose: boolean;
    profile: string | undefined;
    credentialsProcessOutput: boolean;
    force: boolean;
    skipExpiryCheck: boolean;
}

type LogFn = (msg: string) => void;

type Logger = {
    info: LogFn;
    warn: LogFn;
};

const getLogger = (verbose: boolean): Logger => ({
    info: (msg: string): void => {
        if (verbose) {
            console.error('INFO:', msg);
        }
    },
    warn: (msg: string): void => {
        console.error('WARN:', msg);
    },
});

let log: Logger;

interface GetRoleContext {
    ssoConfig: SSOConfigOptions;
    ssoLoginContext: SSOLoginContext;
}

const runGetRoleCredentialsCmdOutput = async (context: GetRoleContext): Promise<string> => {
    if (typeof context.ssoLoginContext.latestCacheFile === 'undefined') {
        throw new NoCachedCredentialsError('Unable to retrieve credentials from SSO cache');
    }

    try {
        const getRoleCredentialsCmdOutput = await execPromise(
            `aws sso get-role-credentials --role-name "${context.ssoConfig.roleName}" --account-id "${context.ssoConfig.accountId}" --access-token "${context.ssoLoginContext.latestCacheFile.accessToken}" --region "${context.ssoLoginContext.latestCacheFile.region}"`,
            {
                env: {
                    ...process.env,
                    AWS_PROFILE: context.ssoLoginContext.awsProfile || 'default',
                },
            },
        );
        return getRoleCredentialsCmdOutput.stdout;
    } catch (err) {
        if (
            hasKey('stderr', err) &&
            err.stderr &&
            typeof err.stderr === 'string' &&
            err.stderr.trim() === 'An error occurred (UnauthorizedException) when calling the GetRoleCredentials operation: Session token not found or invalid'
        ) {
            throw new MisbehavingExpiryDateError('Expiry date is in the future, but the credentials appear to have expired');
        }
        throw err;
    }
};

interface SSOLoginContext {
    haveRunSsoLogin: boolean;
    latestCacheFile: SSOCacheToken | undefined;
    awsProfile: string | undefined;
}

const runSsoLogin = async (context: SSOLoginContext): Promise<void> => {
    log.info('No valid SSO cache file found, running login...');
    const child = spawn('aws', ['sso', 'login'], {
        env: {
            ...process.env,
            AWS_PROFILE: context.awsProfile || 'default',
        },
        stdio: 'inherit',
    });

    await new Promise<void>((res, rej) => {
        child.on('close', (code: number, signal: string) => {
            /* istanbul ignore else */
            if (code === 0) {
                res();
            } else {
                rej(`Command failed with code ${code}, signal: ${signal}`);
            }
        });
    });

    context.haveRunSsoLogin = true;

    log.info('Login completed, trying again to retrieve credentials from SSO cache');
    context.latestCacheFile = await findLatestCacheFile();
};

export const run = async (props: RunProps): Promise<void> => {
    log = getLogger(props.verbose);

    log.info('Starting');

    log.info(`Application version: ${getVersionNumber()}`);

    log.info('Checking CLI version...');

    if (!(await checkCLIVersion())) {
        throw new BadAWSCLIVersionError('Need CLI version 2, see https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html, run `aws --version` to inspect version');
    }

    log.info('CLI version OK');

    if (props.force) {
        log.info('Deleting credentials, and SSO and CLI caches');
        await deleteCredentialsAndCaches();
    }

    if (props.credentialsProcessOutput) {
        log.info('Attempting to retrieve credentials from role credentials cache file');
        const creds = await readCredentialsCacheFile();
        if (typeof creds !== 'undefined') {
            log.info('Found credentials in role credentials cache file, outputting');
            printCredentials(creds);
            return;
        }
        log.info('No valid credentials found in role credentials cache file, continuing');
    }

    log.info('Locating latest SSO cache file...');

    const ssoLoginContext: SSOLoginContext = {
        haveRunSsoLogin: false,
        latestCacheFile: await findLatestCacheFile(),
        awsProfile: props.profile,
    };

    if (props.skipExpiryCheck || typeof ssoLoginContext.latestCacheFile === 'undefined' || ssoLoginContext.latestCacheFile.expiresAt.getTime() < new Date().getTime()) {
        await runSsoLogin(ssoLoginContext);
    }

    log.info('Retrieving SSO configuration from AWS config file...');
    const ssoConfig = await findSSOConfigFromAWSConfig(props.profile);
    log.info('Got SSO configuration');

    log.info('Using SSO credentials to get role credentials...');

    let roleCredentialsOutput: string;
    const getRoleContext = {
        ssoConfig,
        ssoLoginContext,
    };
    try {
        roleCredentialsOutput = await runGetRoleCredentialsCmdOutput(getRoleContext);
    } catch (err) {
        if (err instanceof MisbehavingExpiryDateError) {
            if (ssoLoginContext.haveRunSsoLogin) {
                throw err;
            } else {
                log.info('Expiry date appears to be deceptive, running login...');
                await runSsoLogin(ssoLoginContext);
                log.info('Again using SSO credentials to get role credentials...');
                roleCredentialsOutput = await runGetRoleCredentialsCmdOutput(getRoleContext);
            }
        } else {
            throw err;
        }
    }

    log.info('Parsing role credentials...');
    const roleCredentials = parseRoleCredentialsOutput(roleCredentialsOutput);
    log.info('Got role credentials');

    if (props.credentialsProcessOutput) {
        log.info('Writing role credentials to cache in home directory...');
        await writeCredentialsCacheFile(roleCredentials);
        log.info('Wrote role credentials');

        log.info('Printing role credentials for credentials_process');
        printCredentials(roleCredentials);
        log.info('Printed role credentials');
    } else {
        log.info('Writing role credentials to AWS credentials file...');
        await writeCredentialsFile(roleCredentials, props.profile);
        log.info('Wrote role credentials');
    }

    /* istanbul ignore else */
    if (ssoLoginContext.latestCacheFile) {
        const ssoExpiryDate = ssoLoginContext.latestCacheFile.expiresAt;
        if (ssoExpiryDate) {
            const credentialsExpireInSeconds = (ssoExpiryDate.getTime() - new Date().getTime()) / 1000;
            log.info(`Credentials from SSO expire in ${credentialsExpireInSeconds} seconds`);

            if (credentialsExpireInSeconds < jsSdkExpiryWindowSeconds) {
                log.warn(`Credentials from SSO expire in ${credentialsExpireInSeconds} seconds, which is within the AWS JS SDK's expiry window of ${jsSdkExpiryWindowSeconds} seconds.`);
                log.warn('This may cause issues when using these credentials with the JS SDK, in particular the AWS CDK.');
                log.warn(`Workaround: go to ${ssoConfig.startUrl ?? 'your AWS SSO login URL'}, click 'Sign out', then re-run this command with --force`);
            }
        }
    }

    log.info('Done, exiting cleanly');
};

export const main = async (args: Array<string>): Promise<void> => {
    const parsedArgs = yargs
        .boolean('verbose')
        .describe('verbose', 'Be verbose')
        .string('profile')
        .describe('profile', 'Specify an AWS profile (in ~/.aws/config) to use instead of the default profile')
        .boolean('force')
        .describe('force', 'Delete all credential and cache files before authenticating')
        .boolean('skip-expiry-check')
        .describe('skip-expiry-check', 'Do not check expiry of existing credentials, always reauthenticate')
        .usage('Usage: $0 [...options]')
        .strictOptions()
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        .fail((msg: string, err: Error | null, yargsObj: yargs.Argv): void => {
            /* istanbul ignore next */
            if (err) throw err;
            throw new ArgumentsError(msg);
        })
        .parseSync(args);

    const positionalArgs: Array<string> = parsedArgs._.map((arg: string | number): string => `${arg}`);

    if (positionalArgs.length > 2) {
        throw new ArgumentsError('Too many positional arguments');
    }
    const commands = ['credentials-process', 'version'];

    if (positionalArgs.length > 0 && !commands.includes(positionalArgs[0])) {
        throw new ArgumentsError(`Unexpected argument, expected one of ${commands.join(',')}`);
    }

    if (positionalArgs[0] === 'version') {
        console.log(getVersionNumber());
        return;
    }

    await run({
        verbose: parsedArgs.verbose || false,
        profile: parsedArgs.profile || undefined,
        skipExpiryCheck: !!parsedArgs.skipExpiryCheck,
        force: !!parsedArgs.force,
        credentialsProcessOutput: positionalArgs[0] === 'credentials-process',
    });
};
